packages/api-client-core/src/types.ts

```diff
@@ -36,26 +36,35 @@ export type NonNeverKeys<Selection> = {
  */
 export type FilterNever<T extends Record<string, unknown>> = NonNeverKeys<T> extends never ? never : { [Key in NonNeverKeys<T>]: T[Key] };
 
-/**
- * Extract a subset of a schema given a selection
- *
- * ```typescript
- * type Selection = Select<
- *  { apple: "red", banana: "yellow", orange: "orange" },
- *  { apple: true, banana: false }
- * >;  // { apple: "red" }
- * ```
- */
-export type Select<Schema, Selection extends FieldSelection | null | undefined> = Selection extends null | undefined
+type InnerSelect<Schema, Selection extends FieldSelection | null | undefined> = Selection extends null | undefined
   ? never
   : Schema extends (infer T)[]
   ? Select<T, Selection>[]
   : Schema extends null
   ? Select<Exclude<Schema, null>, Selection> | null
-  : FilterNever<{
+  : {
       [Key in keyof Selection & keyof Schema]: Selection[Key] extends true
         ? Schema[Key]
         : Selection[Key] extends FieldSelection
         ? Select<Schema[Key], Selection[Key]>
         : never;
-    }>;
+    };
+
+/**
+ * Filter out any keys in `T` that are mapped to `never` recursively.
```

Comment:
Might be worth expanding a bit by calling out that objects, after having their `never`-valued keys filtered out, also get filtered out of their parent if they are empty. i.e.,

```ts
type A = DeepFilterNever<{ a: { b: never }, c: "test" }>;
type B = { c: "test" };  // same as A
type C = { a: {}, c: "test" };  // not this one though
```

Suggesting that because an empty object is distinct from `never`. Worth a double check that it's doing this.

Reply:
Added an example in the comment and a test to make sure it's the case.